src/target/at32f43x.c

```diff
@@ -0,0 +1,376 @@
+	target_mem_write32(target, AT32F435_FLASH_STS + bank_reg_offset, status | FLASH_STS_ODF); /* ODF is W1C */
+}
+
+static bool at32f43_flash_busy_wait(
+	target_s *const target, const uint32_t bank_reg_offset, platform_timeout_s *const timeout)
+{
+	/* Read FLASH_STS to poll for Operation Busy Flag */
+	uint32_t status = FLASH_STS_OBF;
+	/* Checking for ODF/EOP requires methodically clearing the ODF */
+	while (!(status & FLASH_STS_ODF) && (status & FLASH_STS_OBF)) {
+		status = target_mem_read32(target, AT32F435_FLASH_STS + bank_reg_offset);
+		if (target_check_error(target)) {
+			DEBUG_ERROR("Lost communications with target\n");
+			return false;
+		}
+		if (timeout)
+			target_print_progress(timeout);
+	}
+	if (status & FLASH_STS_PRGMERR) {
+		DEBUG_ERROR("at32f43 flash error, STS: 0x%" PRIx32 "\n", status);
+		return false;
+	}
+	return !(status & FLASH_STS_PRGMERR);
+}
+
+static inline uint32_t at32f43_bank_offset_for(const target_addr_t addr, const target_addr_t bank_split)
+{
+	if (addr >= bank_split)
+		return FLASH_BANK2_REG_OFFSET;
+	return FLASH_BANK1_REG_OFFSET;
+}
+
+static bool at32f43_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len)
+{
+	target_s *target = flash->t;
+	const at32f43_flash_s *const sf = (at32f43_flash_s *)flash;
+
+	/* Erase range begins in bank 1? Unlock bank 1 */
+	if (addr < sf->bank_split && !at32f43_flash_unlock(target, 0))
+		return false;
+	/* Erase range ends in bank 2? Unlock bank 2 */
+	const target_addr_t end = addr + len - 1U;
+	if (end >= sf->bank_split && !at32f43_flash_unlock(target, FLASH_BANK2_REG_OFFSET))
+		return false;
+
+	for (size_t offset = 0; offset < len; offset += flash->blocksize) {
+		const uint32_t bank_reg_offset = at32f43_bank_offset_for(addr + offset, sf->bank_split);
+		at32f43_flash_clear_eop(target, bank_reg_offset);
+
+		/* Prepare for page/sector erase */
+		target_mem_write32(target, AT32F435_FLASH_CTRL + bank_reg_offset, FLASH_CTRL_SECERS);
+		/* Select erased sector by its address */
+		target_mem_write32(target, AT32F435_FLASH_ADDR + bank_reg_offset, addr + offset);
+		/* Start sector erase operation */
+		target_mem_write32(target, AT32F435_FLASH_CTRL + bank_reg_offset, FLASH_CTRL_SECERS | FLASH_CTRL_ERSTR);
+
+		/* Datasheet: page erase takes 50ms (typ), 500ms (max) */
+		if (!at32f43_flash_busy_wait(target, bank_reg_offset, NULL))
+			return false;
+	}
+	return true;
+}
+
+static inline size_t at32f43_range_in_bank1(const target_addr_t addr, const size_t len, const target_addr_t bank_split)
+{
+	if (addr >= bank_split)
+		return 0;
+	if (addr + len > bank_split)
+		return bank_split - addr;
+	return len;
+}
+
+static bool at32f43_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len)
+{
+	target_s *target = flash->t;
+	const at32f43_flash_s *const sf = (at32f43_flash_s *)flash;
+	const size_t range = at32f43_range_in_bank1(dest, len, sf->bank_split);
```

Comment:
this variable name leaves a bit to be desired

Reply:
It was named "offset", and I decided to rename it to something but didn't want to name it e.g. "length_to_write_in_bank1" which is more descriptive but convoluted.

Reply:
maybe `bank1_range` or `bank1_length`

Reply:
Dropped cross-bank write logic.